### PR TITLE
ROC-7427: Caseworker extending defendant response time

### PR DESCRIPTION
**Pull request by dharmendrak**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ROC-7427

### Change description ###

ROC-7427: Caseworker extending defendant response time

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java [hunk 1]
```diff
@@ -82,83 +71,40 @@ public CallbackResponse prepopulateData(String authorisation) {
...
+    public String generateLetter(CCDCase ccdCase, String authorisation, String templateId)
+        throws DocumentGenerationFailedException {
+        logger.info("General Letter: creating general letter");
+        var docAssemblyResponse = docAssemblyService.createGeneralLetter(ccdCase, authorisation, templateId);
+        return docAssemblyResponse.getRenditionOutputLocation();
     }
 
-    public CallbackResponse printAndUpdateCaseDocuments(CaseDetails caseDetails, String authorisation) {
+    public CCDCase processDocuments(CCDCase ccdCase, Claim claim, String authorisation, String documentName) {
+        var draftLetterDoc = ccdCase.getDraftLetterDoc();
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java:82]: This should probably have a named type as its not immediately obvious that `draftLetterDoc` is of type `CCDDocument`

File: src/main/java/uk/gov/hmcts/cmc/claimstore/rules/MoreTimeRequestRule.java [hunk 2]
```diff
@@ -48,11 +36,6 @@ public void assertMoreTimeCanBeRequested(Claim claim) {
         if (claim.isMoreTimeRequested()) {
             validationErrors.add(ALREADY_REQUESTED_MORE_TIME_ERROR);
         }
-
-        if (claimDeadlineService.isPastDeadline(nowInLocalZone(), newDeadline)) {
-            validationErrors.add(PAST_DEADLINE_ERROR);
-        }
-
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/rules/MoreTimeRequestRule.java:55]: Why are we no longer validating that the new deadline is past the deadline to extend the deadline?

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/CallbackHandlerFactoryTest.java [hunk 3]
```diff
@@ -80,7 +80,7 @@ public void shouldDispatchCallbackForMoreTimeRequested() {
 
         CallbackRequest callbackRequest = CallbackRequest
             .builder()
-            .eventId(MORE_TIME_REQUESTED_PAPER.getValue())
+            .eventId(RESPONSE_MORE_TIME.getValue())
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/CallbackHandlerFactoryTest.java:83]: Should this have changed? Don't we have 2 events, one for citizen requesting more time and one for caseworker?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/rules/MoreTimeRequestRule.java (hunk 2 above)
Comment (Valkrys) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/rules/MoreTimeRequestRule.java:55]: Brian said that this is how the business wants it to be.

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/rules/MoreTimeRequestRule.java:55]: but that was existing feature for frontend, this change is only for caseworker

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/CallbackHandlerFactoryTest.java (hunk 3 above)
Comment (dharmendrak) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/services/CallbackHandlerFactoryTest.java:83]: I feel MORE_TIME_REQUESTED_PAPER and RESPONSE_MORE_TIME are same thing, case worker will receive paper or call too update the response deadline, however would be nice to get clarification on it

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/MoreTimeRequestedCallbackHandler.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        LocalDate newDeadline = responseDeadlineCalculator.calculatePostponedResponseDeadline(claim.getIssuedOn());
+
+        List<String> validationResult = this.moreTimeRequestRule.validateMoreTimeCanBeRequested(claim);
+        var builder = AboutToStartOrSubmitCallbackResponse.builder();
+        if (!validationResult.isEmpty()) {
+            return builder.errors(validationResult).build();
+        }
+
+        return builder
+            .data(Map.of(CALCULATED_RESPONSE_DEADLINE, newDeadline,
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/MoreTimeRequestedCallbackHandler.java:145]: didn't think of, will give a try

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactLetterService.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public String createGeneralLetter(CCDCase ccdCase, String authorisation) {
+        DocAssemblyResponse docAssemblyResponse
+            = docAssemblyService.changeContactLetter(ccdCase, authorisation, generalLetterTemplateId);
+        
+        return docAssemblyResponse.getRenditionOutputLocation();
+    }
+
+    public CallbackResponse printAndUpdateCaseDocuments(CCDCase ccdCase, Claim claim, String authorisation) {
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactLetterService.java]: done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/MoreTimeRequestedCallbackHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/MoreTimeRequestedCallbackHandler.java]: not possible as of design, to directly store we have to show complete defendants

File: src/main/java/uk/gov/hmcts/cmc/claimstore/rules/MoreTimeRequestRule.java (hunk 2 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/rules/MoreTimeRequestRule.java]: reverted

**Pull request merged**